src/Protocol/RecipeMapper.cpp

```diff
@@ -0,0 +1,103 @@
+#include "../Globals.h"
+#include "RecipeMapper.h"
+#include "../Root.h"
+
+cRecipeMapper::cRecipeMapper(void)
+{
+	cFile f;
+	if (!f.Open("Protocol/1.12.2/base.recipes.txt", cFile::fmRead))
+	{
+		LOGWARNING("Cannot open file \"Protocol/1.12.2/base.recipes.txt\", no recipe book recipes will be available!");
+		return;
+	}
+	AString Everything;
+	if (!f.ReadRestOfFile(Everything))
+	{
+		LOGWARNING("Cannot read file \"Protocol/1.12.2/base.recipes.txt\", no recipe book recipes will be available!");
+		return;
+	}
+	f.Close();
+
+	// Split it into lines, then process each line as a single recipe:
+	AStringVector Split = StringSplit(Everything, "\n");
+	m_ProtocolVersionMap["1.12.2"] = {};
+	int LineNum = 1;
+	for (AStringVector::const_iterator itr = Split.begin(); itr != Split.end(); ++itr, ++LineNum)
+	{
+		// Remove anything after a '#' sign and trim away the whitespace:
+		AString Recipe = TrimString(itr->substr(0, itr->find('#')));
+		if (Recipe.empty())
+		{
+			// Empty recipe
+			continue;
+		}
+		AddRecipeLine("1.12.2", LineNum, Recipe);
+	}
+	LOG("Loaded 1.12.2 %zu recipe book", m_ProtocolVersionMap["1.12.2"].size(), LineNum);
+}
+
+
+
+
+
+cRecipeMapper::~cRecipeMapper()
+{
+}
+
+
+
+
+
+void cRecipeMapper::AddRecipeLine(AString a_ProtocolVersion, int a_LineNum, const AString & a_RecipeLine)
+{
+	AString RecipeLine(a_RecipeLine);
```

Comment:
Why take an unnecessary copy?

Reply:
Made it similar to the `CraftingRecipes.cpp`, maybe didn't took enough care if it is necessary.